### ISSUE TITLE
Salt-Master not responding when the list of minions are high after upgrade to 2016.11.2

### ISSUE THREAD

**Issue by daswathn** [saltstack/salt#39863]
### Description of Issue/Question
Have upgraded from _2016.3.2_ to _2016.11.2_ from then on when we run jobs or state_file for minion list which are more than 500 the salt-master stops to respond with the below error. The issue is not just on one salt-master but on all three master servers. 

> 'Salt request timed out. The master is not responding. If this error persists after verifying the master is up, worker_threads may need to be increased.'

**But all these same jobs used to work seamless when we had _2016.3.2_ in the environment**. 

### Setup
Master config file for worker_threads, initially it was set at 50 in version _2016.3.2_ , we increased it to 70 in 2016.11.2, but still the issue persists.
```
salt02:~ # cat /etc/salt/master | grep worker
worker_threads: 70
```

### Steps to Reproduce Issue
```
salt02:~ # salt -N DEW1 test.version  **{DEW1 node group have more than 1200 minion}**
{'DEW001': '2016.11.2'}
{'DEW733': '2016.11.2'}
{'DEW009': '2016.11.2'}
........ Output Truncated 
{'DEW983': '2016.11.2'}
{'DEW998': '2016.11.2'}
{'DEW199': '2016.11.2'}
{'DEW836': '2016.11.2'}
'Salt request timed out. The master is not responding. If this error persists after verifying the master is up, worker_threads may need to be increased.'
```

### Versions Report
```
salt02:~ # salt --versions-report
Salt Version:
           Salt: 2016.11.2

Dependency Versions:
           cffi: 1.1.0
       cherrypy: 3.6.0
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.9 (default, Dec 21 2014, 11:02:59) [GCC]
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.3
            ZMQ: 4.0.4

System Versions:
           dist: SuSE 12 x86_64
        machine: x86_64
        release: 3.12.69-60.64.32-default
         system: Linux
        version: SUSE Linux Enterprise Server  12 x86_64
```

**Comment by Ch3LL** [comment]
@daswathn when you receive the error message is there any significant load on the VM? 

Also is there anything in the debug logs that relates? Even on the minion side as well?

**Comment by daswathn** [comment]
@Ch3LL , Yes your right the load increases signiciently on the VM , but only one salt-worker logs the load on the server with, when we did a strace on the PID 

```
Command that we ran on all the hosts 
salt02:~ # salt \* cmd.run "/snt/prod/data/Gaur/rsh-disable.pl"
```

Strace on the process salt-master process which is showing the below trace..  
```
salt02:~ # strace -p 17216
Process 17216 attached
stat("/etc/salt/pki/master/minions/DFGLD05450", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05451", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05452", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05453", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05465", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05466", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05467", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05471", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05472", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05473", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05474", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05475", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05476", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05477", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05478", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05480", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05481", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05483", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05484", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05485", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05486", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
stat("/etc/salt/pki/master/minions/DFGLD05487", {st_mode=S_IFREG|0644, st_size=450, ...}) = 0
```

The logs are written as below, for all the 8000+ when we run the across the landscape..

```
2017-03-09 06:25:43,621 [salt.utils.event ][DEBUG   ][18632] Sending event: tag = salt/job/20170309062538899739/ret/ld7325; data = {'fun_args': ['20170309061843871420'], 'jid': '20170309062538899739', 'return': {}, 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2017-03-09T05:25:43.621167', 'fun': 'saltutil.find_job', 'id': 'ld7325'}
2017-03-09 06:25:43,621 [salt.utils.job   ][INFO    ][18524] Got return from DFGLD05119 for job 20170309062538899739
2017-03-09 06:25:43,621 [salt.utils.event ][DEBUG   ][18575] Sending event: tag = salt/job/20170309062538899739/ret/DFGLD03229; data = {'fun_args': ['20170309061843871420'], 'jid': '20170309062538899739', 'return': {}, 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2017-03-09T05:25:43.621373', 'fun': 'saltutil.find_job', 'id': 'DFGLD03229'}
```

We will also look for more clues, but if your already aware of this issues, kindly let us know where we can fix this.

**Comment by ramachandranviji** [comment]
We are having a similar issue with our landscape where we installed salt 11.1 and then upgraded to salt 2016.11.2. But as we run some jobs across a list of servers more than 300, it throws the below error message. 
'Salt request timed out. The master is not responding. If this error persists after verifying the master is up, worker_threads may need to be increased.'

This makes the work really difficult as ours is a big landscape (8500+ servers) and we can't work on small sets of servers as it will take forever to complete a job. While checking the processes, what we find a 'salt-master' process that is continuously checking the keys and one other 'salt' process on strace of which we are receiving timeouts and resources unavailable issues.

**Comment by Ch3LL** [comment]
I will have an update on my test case tomorrow morning. in teh meantime @ramachandranviji what OS are you seeing this on? 

Also both @daswathn or @ramachandranviji do you upgrade your minions to 2016.11 or are they on a lower version then the master? Just need specifics for the test case thanks

**Comment by ramachandranviji** [comment]
OS release - SUSE Linux Enterprise Server 12 (x86_64), VERSION = 12, PATCHLEVEL = 1
The minions and the master are on the same version salt-call 2016.11.2 (Carbon)

**Comment by Ch3LL** [comment]
Okay i'm NOT able to replicate this error: `'Salt request timed out. The master is not responding. If this error persists after verifying the master is up, worker_threads may need to be increased.'` but I am seeing a big CPU spike in between versions 2016.3.2 and 2016.11.2/2016.11.3.  

In 2016.3.4 the CPU spiked around 34% on the salt-master and salt process throughout the execution of ` salt '*' test.ping` against 600 hosts. In 2016.11.2/3 I am seeing the CPU spike anywhere between 40-99% in top during the execution process against 600 hosts. So I am guessing maybe I just have more CPU capabilities on my host and thats why i'm not seeing the salt request timed out error on the master. 

Is it your CPU that is mostly spiking on your load tests? My memory didn't have much load.

**Comment by ramachandranviji** [comment]
Yes, in my case its the CPU that is spiking mostly. The memory was not loaded. Is there any suggestions on how to get this fixed or any possible workarounds?

**Comment by Ch3LL** [comment]
I'm not aware of any workaround other then to say if your CPU is the spike if its easy to add more CPU to see if you don't get the error still. Or possibly work around with the number of worker_threads you have set in your master config file, but seems that might not resolve the issue. Maybe @cachedout has some better ideas on a workaround?

One another note this is a critical issue for us and is a release blocker for us. This will soon get assigned out to an engineer and looked at to get fixed. And he/she will have better suggestions for you and a fix.

**Comment by cachedout** [comment]
User-space CPU or is this just IO wait?

**Comment by rallytime** [comment]
Hi @daswathn - I have fixed this issue with the PR listed above, #40615. That fix will be available in the upcoming 2016.11.4 release.
